client/my-sites/promote-post-i2/index.js

```diff
@@ -18,14 +20,28 @@ export const redirectToPrimarySite = ( context, next ) => {
 		return next();
 	}
 
-	const state = context.store.getState();
-	const primarySiteSlug = getPrimarySiteSlug( state );
-	if ( primarySiteSlug !== null ) {
-		page( `${ context.pathname }/${ primarySiteSlug }` );
-	} else {
-		siteSelection( context, next );
-		page( getAdvertisingDashboardPath( '' ) );
+	const { getState, dispatch } = context.store;
+	const primarySiteId = getPrimarySiteId( getState() );
+	const primarySiteSlug = getSiteSlug( getState(), primarySiteId );
+
+	if ( primarySiteSlug ) {
+		redirectToPrimary( context, primarySiteSlug );
+		return;
 	}
+
+	// Fetch the primary site by ID and then try to determine its slug again.
+	dispatch( requestSite( primarySiteId ) )
+		.catch( () => null )
+		.then( () => {
+			const freshPrimarySiteSlug = getSiteSlug( getState(), primarySiteId );
+			if ( freshPrimarySiteSlug ) {
+				redirectToPrimary( context, freshPrimarySiteSlug );
+				return;
+			}
+
+			// no redirection happened, proceed to showing the sites list
+			next();
+		} );
```

Comment:
Would be nice if we could reuse the part that is the same as in `siteSelection()` instead of having to duplicate it.

Reply:
Yes, unfortunately the relevant code in `siteSelection` runs only when `hasOneSite` is true. But we could extract it into a helper, just like `redirectToPrimary`.